velox/serializers/RowSerializer.h

```diff
@@ -0,0 +1,190 @@
+#include "velox/vector/ComplexVector.h"
+#include "velox/vector/VectorStream.h"
+
+namespace facebook::velox::serializer {
+
+using TRowSize = uint32_t;
+
+template <class Serializer>
+class RowSerializer : public IterativeVectorSerializer {
+ public:
+  explicit RowSerializer(StreamArena* streamArena)
+      : pool_{streamArena->pool()} {}
+
+  void append(
+      const RowVectorPtr& vector,
+      const folly::Range<const IndexRange*>& ranges,
+      Scratch& /*scratch*/) override {
+    size_t totalSize = 0;
+    const auto totalRows = std::accumulate(
+        ranges.begin(),
+        ranges.end(),
+        0,
+        [](vector_size_t sum, const auto& range) { return sum + range.size; });
+
+    if (totalRows == 0) {
+      return;
+    }
+
+    Serializer row(vector);
+    std::vector<vector_size_t> rowSize(totalRows);
+    if (auto fixedRowSize =
+            Serializer::fixedRowSize(asRowType(vector->type()))) {
+      totalSize += (fixedRowSize.value() + sizeof(TRowSize)) * totalRows;
+      std::fill(rowSize.begin(), rowSize.end(), fixedRowSize.value());
+    } else {
+      vector_size_t index = 0;
+      for (const auto& range : ranges) {
+        for (auto i = 0; i < range.size; ++i, ++index) {
+          rowSize[index] = row.rowSize(range.begin + i);
+          totalSize += rowSize[index] + sizeof(TRowSize);
+        }
+      }
+    }
+
+    if (totalSize == 0) {
+      return;
+    }
+
+    BufferPtr buffer = AlignedBuffer::allocate<char>(totalSize, pool_, 0);
+    auto* rawBuffer = buffer->asMutable<char>();
+    buffers_.push_back(std::move(buffer));
+
+    serializeRanges(row, ranges, rawBuffer, rowSize);
+  }
+
+  void append(
+      const Serializer& compactRow,
+      const folly::Range<const vector_size_t*>& rows,
+      const std::vector<vector_size_t>& sizes) override {
+    size_t totalSize = 0;
+    for (const auto row : rows) {
+      totalSize += sizes[row];
+    }
+    if (totalSize == 0) {
+      return;
+    }
+
+    BufferPtr buffer = AlignedBuffer::allocate<char>(totalSize, pool_, 0);
+    auto* rawBuffer = buffer->asMutable<char>();
+    buffers_.push_back(std::move(buffer));
+
+    size_t offset = 0;
+    for (auto& row : rows) {
+      // Write row data.
+      const TRowSize size =
+          compactRow.serialize(row, rawBuffer + offset + sizeof(TRowSize));
+
+      // Write raw size. Needs to be in big endian order.
+      *(TRowSize*)(rawBuffer + offset) = folly::Endian::big(size);
+      offset += sizeof(TRowSize) + size;
+    }
+  }
+
+  size_t maxSerializedSize() const override {
+    size_t totalSize = 0;
+    for (const auto& buffer : buffers_) {
+      totalSize += buffer->size();
+    }
+    return totalSize;
+  }
+
+  void flush(OutputStream* stream) override {
+    for (const auto& buffer : buffers_) {
+      stream->write(buffer->as<char>(), buffer->size());
+    }
+    buffers_.clear();
+  }
+
+  void clear() override {}
+
+ protected:
+  virtual void serializeRanges(
```

Comment:
After this PR, we can unify serializeRanges https://github.com/facebookincubator/velox/pull/11206